components/site/src/lib.rs

```diff
@@ -583,11 +584,29 @@ impl Site {
         imageproc.do_process()
     }
 
-    /// Deletes the `public` directory if it exists
+    /// Deletes the contents in the `public` directory if it exists
     pub fn clean(&self) -> Result<()> {
         if self.output_path.exists() {
-            // Delete current `public` directory so we can start fresh
-            remove_dir_all(&self.output_path).context("Couldn't delete output directory")?;
+            // Delete contents of current `public` directory so we can start fresh
+            for entry in self.output_path.read_dir().context(format!(
+                "Couldn't read output directory `{}`",
+                self.output_path.display()
+            ))? {
+                let entry = entry.context("Couldn't read entry in output directory")?.path();
+
+                // Skip hidden files and folders if the preserve_hidden_file configuration option is set
+                if self.config.preserve_hidden_files && file_hidden(&entry) {
+                    continue;
+                }
+
+                if entry.is_dir() {
+                    remove_dir_all(entry)
+                        .context("Couldn't delete folder while cleaning the output directory")?;
+                } else {
+                    remove_file(entry)
+                        .context("Couldn't delete file while cleaning the output directory")?;
+                }
+            }
```

Comment:
That doesn't work recursively so it only preserves hidden files at the root of the dir?

Reply:
Should it work recursively? It would result that regular directories are kept when they contain any hidden file inside them so it could potentially result in odd looking behaviour from a user perspective.
My personal intuition would be that this configuration option only affects files and folders at the root level. I can make that more clear in the docs

Reply:
I'm wondering if root should appear in the config name to be more explicit

Reply:
I was thinking about it too but I could not come up with a concise name

Reply:
Let's think about it for a few days and see if anything comes up

Reply:
How about `preserve_dotfiles_in_public`